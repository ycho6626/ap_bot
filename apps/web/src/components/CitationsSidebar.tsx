import { X, ExternalLink, BookOpen } from 'lucide-react';
import { cn } from '@/lib/utils';

interface Source {
  type: 'canonical' | 'retrieval' | 'generated';
  id: string;
  title: string;
  snippet: string;
  score: number;
}

interface CitationsSidebarProps {
  sources: Source[];
  suggestions?: string[];
  onClose: () => void;
  className?: string;
}

export function CitationsSidebar({ 
  sources, 
  suggestions = [], 
  onClose, 
  className 
}: CitationsSidebarProps) {
  const getSourceTypeColor = (type: Source['type']) => {
    switch (type) {
      case 'canonical':
        return 'bg-success-100 text-success-800';
      case 'retrieval':
        return 'bg-primary-100 text-primary-800';
      case 'generated':
        return 'bg-warning-100 text-warning-800';
      default:
        return 'bg-gray-100 text-gray-800';
    }
  };

  const getSourceTypeLabel = (type: Source['type']) => {
    switch (type) {
      case 'canonical':
        return 'Canonical Solution';
      case 'retrieval':
        return 'Knowledge Base';
      case 'generated':
        return 'Generated';
      default:
        return 'Unknown';
    }
  };

  return (
    <div className={cn('w-80 bg-white border border-gray-200 rounded-lg shadow-lg', className)}>
      {/* Header */}
      <div className="flex items-center justify-between p-4 border-b border-gray-200">
        <div className="flex items-center space-x-2">
          <BookOpen className="h-5 w-5 text-gray-600" />
          <h3 className="text-lg font-semibold text-gray-900">Sources & Citations</h3>
        </div>
        <button
          onClick={onClose}
          className="p-1 hover:bg-gray-100 rounded-md transition-colors"
          aria-label="Close citations sidebar"
        >
          <X className="h-4 w-4 text-gray-500" />
        </button>
      </div>

      {/* Content */}
      <div className="p-4 space-y-6 max-h-[calc(100vh-200px)] overflow-y-auto">
        {/* Sources */}
        {sources.length > 0 && (
          <div>
            <h4 className="text-sm font-medium text-gray-900 mb-3">Sources Used</h4>
            <div className="space-y-3">
              {sources.map((source, index) => (
                <div
                  key={`${source.id}-${index}`}
                  className="border border-gray-200 rounded-lg p-3 hover:bg-gray-50 transition-colors"
                >
                  <div className="flex items-start justify-between mb-2">
                    <span
                      className={cn(
                        'inline-flex items-center px-2 py-1 rounded-full text-xs font-medium',
                        getSourceTypeColor(source.type)
                      )}
                    >
                      {getSourceTypeLabel(source.type)}
                    </span>
                    <span className="text-xs text-gray-500">
                      {Math.round(source.score * 100)}% match
                    </span>
                  </div>
                  
                  <h5 className="text-sm font-medium text-gray-900 mb-2 line-clamp-2">
                    {source.title}
                  </h5>
                  
                  <p className="text-xs text-gray-600 mb-3 line-clamp-3">
                    {source.snippet}
                  </p>
                  
                  <button
                    onClick={() => {
                      // TODO: Implement source viewing
                      console.log('View source:', source.id);
                    }}
                    className="inline-flex items-center text-xs text-primary-600 hover:text-primary-700 transition-colors"
                  >
                    <ExternalLink className="h-3 w-3 mr-1" />
                    View Source
                  </button>
                </div>
              ))}
            </div>
          </div>
        )}

        {/* Suggestions */}
        {suggestions.length > 0 && (
          <div>
            <h4 className="text-sm font-medium text-gray-900 mb-3">Suggestions</h4>
            <div className="space-y-2">
              {suggestions.map((suggestion, index) => (
                <div
                  key={index}
                  className="flex items-start space-x-2 p-2 bg-gray-50 rounded-md"
                >
                  <div className="w-1.5 h-1.5 bg-primary-500 rounded-full mt-2 flex-shrink-0" />
                  <p className="text-sm text-gray-700">{suggestion}</p>
                </div>
              ))}
            </div>
          </div>
        )}

        {/* Empty State */}
        {sources.length === 0 && suggestions.length === 0 && (
          <div className="text-center py-8">
            <BookOpen className="h-8 w-8 text-gray-400 mx-auto mb-3" />
            <p className="text-sm text-gray-500">
              No sources or suggestions available for this response.
            </p>
          </div>
        )}
      </div>
    </div>
  );
}
