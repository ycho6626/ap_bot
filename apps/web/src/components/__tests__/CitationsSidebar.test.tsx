import { describe, it, expect, vi } from 'vitest';
import { render, screen, fireEvent } from '@testing-library/react';
import { CitationsSidebar } from '../CitationsSidebar';

const mockSources = [
  {
    type: 'canonical' as const,
    id: 'source-1',
    title: 'Derivative Rules',
    snippet: 'The derivative of x^n is nx^(n-1)',
    score: 0.95
  },
  {
    type: 'retrieval' as const,
    id: 'source-2',
    title: 'Integration Techniques',
    snippet: 'Integration by parts formula...',
    score: 0.87
  }
];

const mockSuggestions = [
  'Try asking about the chain rule',
  'Learn more about definite integrals'
];

describe('CitationsSidebar', () => {
  it('should render sources with correct information', () => {
    render(
      <CitationsSidebar
        sources={mockSources}
        suggestions={mockSuggestions}
        onClose={vi.fn()}
      />
    );

    expect(screen.getByText('Sources & Citations')).toBeInTheDocument();
    expect(screen.getByText('Sources Used')).toBeInTheDocument();
    expect(screen.getByText('Derivative Rules')).toBeInTheDocument();
    expect(screen.getByText('Integration Techniques')).toBeInTheDocument();
  });

  it('should render source type badges correctly', () => {
    render(
      <CitationsSidebar
        sources={mockSources}
        suggestions={[]}
        onClose={vi.fn()}
      />
    );

    expect(screen.getByText('Canonical Solution')).toBeInTheDocument();
    expect(screen.getByText('Knowledge Base')).toBeInTheDocument();
  });

  it('should render source scores correctly', () => {
    render(
      <CitationsSidebar
        sources={mockSources}
        suggestions={[]}
        onClose={vi.fn()}
      />
    );

    expect(screen.getByText('95% match')).toBeInTheDocument();
    expect(screen.getByText('87% match')).toBeInTheDocument();
  });

  it('should render suggestions when provided', () => {
    render(
      <CitationsSidebar
        sources={[]}
        suggestions={mockSuggestions}
        onClose={vi.fn()}
      />
    );

    expect(screen.getByText('Suggestions')).toBeInTheDocument();
    expect(screen.getByText('Try asking about the chain rule')).toBeInTheDocument();
    expect(screen.getByText('Learn more about definite integrals')).toBeInTheDocument();
  });

  it('should render external links with correct attributes', () => {
    render(
      <CitationsSidebar
        sources={mockSources}
        suggestions={[]}
        onClose={vi.fn()}
      />
    );

    const links = screen.getAllByText('View Source');
    expect(links).toHaveLength(2);
    
    links.forEach(link => {
      expect(link).toHaveAttribute('target', '_blank');
      expect(link).toHaveAttribute('rel', 'noopener noreferrer');
    });
  });

  it('should call onClose when close button is clicked', () => {
    const mockOnClose = vi.fn();
    render(
      <CitationsSidebar
        sources={[]}
        suggestions={[]}
        onClose={mockOnClose}
      />
    );

    const closeButton = screen.getByLabelText('Close citations sidebar');
    fireEvent.click(closeButton);
    
    expect(mockOnClose).toHaveBeenCalledTimes(1);
  });

  it('should render empty state when no sources or suggestions', () => {
    render(
      <CitationsSidebar
        sources={[]}
        suggestions={[]}
        onClose={vi.fn()}
      />
    );

    expect(screen.getByText('No sources or suggestions available for this response.')).toBeInTheDocument();
  });

  it('should apply custom className', () => {
    const { container } = render(
      <CitationsSidebar
        sources={[]}
        suggestions={[]}
        onClose={vi.fn()}
        className="custom-class"
      />
    );

    expect(container.firstChild).toHaveClass('custom-class');
  });
});
