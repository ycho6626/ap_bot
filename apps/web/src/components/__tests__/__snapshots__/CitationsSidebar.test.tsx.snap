// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CitationsSidebar > matches snapshot with sources and suggestions 1`] = `
<div
  class="w-80 bg-white border border-gray-200 rounded-lg shadow-lg"
  data-testid="citations-sidebar"
>
  <div
    class="flex items-center justify-between p-4 border-b border-gray-200"
  >
    <div
      class="flex items-center space-x-2"
    >
      <svg
        class="lucide lucide-book-open h-5 w-5 text-gray-600"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M2 3h6a4 4 0 0 1 4 4v14a3 3 0 0 0-3-3H2z"
        />
        <path
          d="M22 3h-6a4 4 0 0 0-4 4v14a3 3 0 0 1 3-3h7z"
        />
      </svg>
      <h3
        class="text-lg font-semibold text-gray-900"
      >
        Sources & Citations
      </h3>
    </div>
    <button
      aria-label="Close citations sidebar"
      class="p-1 hover:bg-gray-100 rounded-md transition-colors"
    >
      <svg
        class="lucide lucide-x h-4 w-4 text-gray-500"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18 6 6 18"
        />
        <path
          d="m6 6 12 12"
        />
      </svg>
    </button>
  </div>
  <div
    class="p-4 space-y-6 max-h-[calc(100vh-200px)] overflow-y-auto"
  >
    <div
      data-testid="citations-sources"
    >
      <h4
        class="text-sm font-medium text-gray-900 mb-3"
      >
        Sources Used
      </h4>
      <div
        class="space-y-3"
      >
        <div
          class="border border-gray-200 rounded-lg p-3 hover:bg-gray-50 transition-colors"
        >
          <div
            class="flex items-start justify-between mb-2"
          >
            <span
              class="inline-flex items-center px-2 py-1 rounded-full text-xs font-medium bg-success-100 text-success-800"
            >
              Canonical Solution
            </span>
            <span
              class="text-xs text-gray-500"
            >
              95
              % match
            </span>
          </div>
          <h5
            class="text-sm font-medium text-gray-900 mb-2 line-clamp-2"
          >
            Derivative Rules
          </h5>
          <p
            class="text-xs text-gray-600 mb-3 line-clamp-3"
          >
            The derivative of x^n is nx^(n-1)
          </p>
          <a
            class="inline-flex items-center text-xs text-primary-600 hover:text-primary-700 transition-colors"
            href="/lessons/source-1"
            rel="noopener noreferrer"
            target="_blank"
          >
            <svg
              class="lucide lucide-external-link h-3 w-3 mr-1"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
              />
              <polyline
                points="15 3 21 3 21 9"
              />
              <line
                x1="10"
                x2="21"
                y1="14"
                y2="3"
              />
            </svg>
            View Source
          </a>
        </div>
        <div
          class="border border-gray-200 rounded-lg p-3 hover:bg-gray-50 transition-colors"
        >
          <div
            class="flex items-start justify-between mb-2"
          >
            <span
              class="inline-flex items-center px-2 py-1 rounded-full text-xs font-medium bg-primary-100 text-primary-800"
            >
              Knowledge Base
            </span>
            <span
              class="text-xs text-gray-500"
            >
              87
              % match
            </span>
          </div>
          <h5
            class="text-sm font-medium text-gray-900 mb-2 line-clamp-2"
          >
            Integration Techniques
          </h5>
          <p
            class="text-xs text-gray-600 mb-3 line-clamp-3"
          >
            Integration by parts formula...
          </p>
          <a
            class="inline-flex items-center text-xs text-primary-600 hover:text-primary-700 transition-colors"
            href="/lessons/source-2"
            rel="noopener noreferrer"
            target="_blank"
          >
            <svg
              class="lucide lucide-external-link h-3 w-3 mr-1"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M18 13v6a2 2 0 0 1-2 2H5a2 2 0 0 1-2-2V8a2 2 0 0 1 2-2h6"
              />
              <polyline
                points="15 3 21 3 21 9"
              />
              <line
                x1="10"
                x2="21"
                y1="14"
                y2="3"
              />
            </svg>
            View Source
          </a>
        </div>
      </div>
    </div>
    <div>
      <h4
        class="text-sm font-medium text-gray-900 mb-3"
      >
        Suggestions
      </h4>
      <div
        class="space-y-2"
      >
        <div
          class="flex items-start space-x-2 p-2 bg-gray-50 rounded-md"
        >
          <div
            class="w-1.5 h-1.5 bg-primary-500 rounded-full mt-2 flex-shrink-0"
          />
          <p
            class="text-sm text-gray-700"
          >
            Try asking about the chain rule
          </p>
        </div>
        <div
          class="flex items-start space-x-2 p-2 bg-gray-50 rounded-md"
        >
          <div
            class="w-1.5 h-1.5 bg-primary-500 rounded-full mt-2 flex-shrink-0"
          />
          <p
            class="text-sm text-gray-700"
          >
            Learn more about definite integrals
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
